src/pages/spawn/spawnHost/SpawnHostModal.tsx

```diff
@@ -141,22 +140,38 @@ export const SpawnHostModal: React.FC<SpawnHostModalProps> = ({
     }
   }, [awsRegions, publicKeys, dispatch]);
 
-  // Need to initialize these here so they can be used in the useEffect hook
-  let virtualWorkstationDistros = [];
-  let notVirtualWorkstationDistros = [];
+  const virtualWorkstationDistros = useMemo(
+    () =>
+      (distrosData?.distros ?? [])
+        .filter((d) => d.name.includes(distroInput))
+        .filter((d) => d.isVirtualWorkStation)
+        .sort((a, b) => a.name.localeCompare(b.name)),
+    [distrosData?.distros, distroInput]
+  );
+  const notVirtualWorkstationDistros = useMemo(
+    () =>
+      (distrosData?.distros ?? [])
+        .filter((d) => d.name.includes(distroInput))
+        .filter((d) => !d.isVirtualWorkStation)
+        .sort((a, b) => a.name.localeCompare(b.name)),
+    [distrosData?.distros, distroInput]
+  );
+
+  // recalculate isVirtualWorkstation whenever distro changes
+  // initial distroId can be changed from URL
+  useEffect(() => {
+    dispatch({
+      type: "editDistroEffect",
+      isVirtualWorkstation: !!virtualWorkstationDistros.find(
+        (vd) => distroId === vd.name
+      ),
+    });
+  }, [distroId, dispatch, virtualWorkstationDistros]);
```

Comment:
I also update the distro from the `ingestQueryParams` action. If I follow the pattern you describe, the component that dispatches the `ingestQueryParams` will also need access to `virtualWorkstationDistros` from useQuery (or readQuery) and will also need to apply this logic again: 
```
    isVirtualWorkstation: !!virtualWorkstationDistros.find(
        (vd) => distroId === vd.name
      ),
```
If I adjust the props to PublicKeyForm to only accept a subset of values for spawnHostModalState I think we should expect very similar dom updates with only 1 additional call to render and we will get the benefit of not having to maintain isVirtualWorkstation across multiple actions. what do you think?

Reply:
@khelif96 I decided not to adjust public key form props since the refactor may require additional testing to ensure we do not run into undefined value errors